### PR TITLE
create proc/func/trigger, create/drop database, transaction statements are not being tracked by the extensions which utilize processutility hook

### PR DESCRIPTION
**Pull request by kritika-0601**
Currently statements like create proc/func/trigger/database, drop database and transaction statements are not being tracked by the extensions which utilises the processutility hooks in Babelfish. Above issue happens because bbf_processutility doesn't call subsequent/previous processutility hook for above statements due to which these subsequent extension hooks are not being executed.
This commit resolves above issue by adding a custom processutility hook in engine to execute above Babelfish statements so that execution happens at last step of processutility hook chain which makes sure that subsequent extension hooks are executed.
* Before:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> executes above statements -> execution completed/doesn't return to previous hook execution or doesn't execute subsequent hooks

* After:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> ... -> processUtility_hooks()-> ... -> Standard_processUtility()-> executes above statements using custom hook -> returns to previous hook execution

Task: BABEL-4218

Link for engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/127

Since this PR includes calling the engine side function to extension side using hook, we do not require extra test scenarios to test it, it is already being covered in previous tests.
We have separate test PR for testing pg_stat_statements extension.
### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 1]
```diff
@@ -364,12 +374,239 @@ UninstallExtendedHooks(void)
...
+	Node	   *parsetree = pstmt->utilityStmt;
+
+
+	pstate->p_sourcetext = queryString;
+
+	switch (nodeTag(parsetree))
+	{
+		case T_CreateFunctionStmt:
+		{
+			CreateFunctionStmt *stmt = (CreateFunctionStmt *)parsetree;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.c]: add implementation for create function into a separate utility function

File: contrib/babelfishpg_tsql/src/hooks.h [hunk 2]
```diff
@@ -22,4 +23,8 @@ extern Oid	get_tsql_trigger_oid(List *object,
 
 extern char *update_delete_target_alias;
 extern bool sp_describe_first_result_set_inprogress;
+
+extern void PLTsqlProcessTransaction(Node *parsetree,
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.h]: this should not be in hooks, move to utility file

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 3]
```diff
@@ -1884,7 +1884,7 @@ bbf_table_var_lookup(const char *relname, Oid relnamespace)
 /*
  * Transaction processing using tsql semantics
  */
-static void
+extern void
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: move it to utility file

File: contrib/babelfishpg_tsql/src/hooks.h [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #include "postgres.h"
 #include "catalog/catalog.h"
 #include "parser/analyze.h"
+#include "tcop/cmdtag.h"
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.h]: why this header? If for pointer types use forwards declaration

File: contrib/babelfishpg_tsql/src/hooks.h [hunk 5]
```diff
@@ -24,4 +24,5 @@ extern Oid	get_tsql_trigger_oid(List *object,
 
 extern char *update_delete_target_alias;
 extern bool sp_describe_first_result_set_inprogress;
+
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.h]: unnecessary change

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 6]
```diff
@@ -67,6 +75,282 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
  * Also, length should be restricted to 8000 for sys.varchar and sys.char datatypes.
  * And length should be restricted to 4000 for sys.varchar and sys.char datatypes
  */
+/*
+ * Transaction processing using tsql semantics
+ */
+extern void
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: why extern here?

Comment (kritika-0601) [reply contrib/babelfishpg_tsql/src/pltsql_utils.c]: Yeah, I will remove it.

**Review by surendravishnoi — DISMISSED** [review]

**Review by surendravishnoi — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 7]
```diff
@@ -51,6 +59,281 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
...
+	}
+}
+
+
+bool
+pltsql_createFunction(ParseState *pstate, PlannedStmt *pstmt, const char *queryString, ProcessUtilityContext context, 
+                          ParamListInfo params)
+{
+	Node	   *parsetree = pstmt->utilityStmt;
+    CreateFunctionStmt *stmt = (CreateFunctionStmt *)parsetree;
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Please fix the indentation.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 8]
```diff
@@ -51,6 +59,281 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
...
+				if (strcmp(defel->defname, "language") == 0)
+				{
+					if (language_item)
+						ereport(ERROR,
+								(errcode(ERRCODE_SYNTAX_ERROR),
+								errmsg("conflicting or redundant options"),
+								parser_errposition(pstate, defel->location)));
+					language_item = defel;
+				}
+			}
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Please fix indentation.

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 9]
```diff
@@ -51,6 +59,281 @@ const uint64 PLTSQL_LOCKTAG_OFFSET = 0xABCDEF;
...
+
+	if (language_item)
+		language = strVal(language_item->arg);
+
+	if(!((language && !strcmp(language,"pltsql")) || sql_dialect == SQL_DIALECT_TSQL))
+	{
+		return false;
+	}
+
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Remove this line.

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

**Review by HarshLunagariya — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]

**Pull request merged**